example/matrixAddWithMdspan/src/matrixAddMdSpan.cpp

```diff
@@ -0,0 +1,166 @@
+/* Copyright 2023 Mehmet Yusufoglu, Bernhard Manfred Gruber, René Widera, Andrea Bocci, Simeon Ehrig
+ * SPDX-License-Identifier: ISC
+ */
+
+#include <alpaka/alpaka.hpp>
+// Needed for running example for all backends available; one by one
+#include <alpaka/example/ExecuteForEachAccTag.hpp>
+
+#include <experimental/mdspan>
```

Comment:
should this be wrapped in a check on the C++ standard, so if we are using c++23 we take `<mdspan>` and otherise if we are using the version that comes with alpaka we take `"experimental/mdspan"` ?

Reply:
I disagree because the version shipped with alpaka is different to the standard implementation in an important point. The alpaka version uses the `operator()` [1] to access elements. The standard implementation uses the `operator[]` because since C++ 23 more than one argument is allowed. Therefore we would need to add a preporcessor if-else or wrapper function each time the memory is accessed.

[1] The experimental implementation provides also the `operator[]` if we use C++ 23. But we also support C++17 and 20.

Reply:
Mhm, I see.

On the other hand, `alpaka` (the library) does not _use_ the `mdspan` objects, it only returns them.

I think that, if somebody uses `alpaka` in a c++23 environment, they should get the official `mdspan`, not the experimental version (like we do for `boost::atomic_ref` vs `std::atomic_ref`).

Can we `#if __cplusplus` the examples to use `()` for c++17/c++20, and `[]` for c++23 ?

Reply:
I agree, but I think this can be a little bit more work.

The implementation we use creates a defines `MDSPAN_USE_BRACKET_OPERATOR`. Than it looks like this: https://github.com/SimeonEhrig/MyCpp/blob/28d20632657697533184f95c883739718f8397ad/features/23/mdspan/hello_mspan/main.cpp#L44
I expect, that the define will be not created by the standard library and we need to create it by our self.

Another point is, that our implementation annotates the functions with `__host__` and `__device__` to make it compatible to CUDA and HIP. At the moment, Nvidia does not provide a stl implementation and I'm not sure, if AMD does it: https://en.cppreference.com/w/cpp/compiler_support#cpp23
But Clang provides already an implementation. Therefore, in the worst case wit detect the stl implementation of Clang, which is maybe not annotated  and tries to compile it with the CUDA or HIP backend. It this case, we need a workaround selection mechanism, if the mdspan implementation is not annotated.

Long story short, I agree that we should be forward compatible but I think we need to solve this in an extra PR.

Reply:
> Long story short, I agree that we should be forward compatible but I think we need to solve this in an extra PR.

OK for me.

Reply:
ok, i will open a related issue not to forget, or @SimeonEhrig  could you open since you know the details...